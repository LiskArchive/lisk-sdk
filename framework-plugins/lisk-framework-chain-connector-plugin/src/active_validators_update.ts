/*
 * Copyright Â© 2022 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
/* eslint-disable no-bitwise */
import { ActiveValidator, utils, ActiveValidatorsUpdate } from 'lisk-sdk';
import { ValidatorsData } from './types';

/**
 * @see https://github.com/LiskHQ/lips/blob/main/proposals/lip-0053.md#computing-the-validators-update
 */

export const calculateActiveValidatorsUpdate = (
	validatorsDataAtLastCertificate: ValidatorsData,
	validatorsDataAtNewCertificate: ValidatorsData,
): { activeValidatorsUpdate: ActiveValidatorsUpdate; certificateThreshold: bigint } => {
	let certificateThreshold;
	// If the certificate threshold is not changed from last certificate then we assign zero
	if (
		validatorsDataAtNewCertificate.certificateThreshold ===
		validatorsDataAtLastCertificate.certificateThreshold
	) {
		certificateThreshold = validatorsDataAtLastCertificate.certificateThreshold;
	} else {
		certificateThreshold = validatorsDataAtNewCertificate.certificateThreshold;
	}

	const activeValidatorsUpdate = getActiveValidatorsUpdate(
		validatorsDataAtLastCertificate.validators,
		validatorsDataAtNewCertificate.validators,
	);

	return { activeValidatorsUpdate, certificateThreshold };
};

/**
 * @see https://github.com/LiskHQ/lips/blob/main/proposals/lip-0053.md#computing-the-validators-update
 */
export const getActiveValidatorsUpdate = (
	currentValidators: ActiveValidator[],
	newValidators: ActiveValidator[],
): ActiveValidatorsUpdate => {
	const currentValidatorsMap = new utils.dataStructures.BufferMap<ActiveValidator>();
	const allBLSKeysSet = new utils.dataStructures.BufferSet();
	for (const v of currentValidators) {
		currentValidatorsMap.set(v.blsKey, v);
		allBLSKeysSet.add(v.blsKey);
	}

	const validatorsUpdate: ActiveValidator[] = [];
	const blsKeysUpdate: Buffer[] = [];
	const newBLSKeys = new utils.dataStructures.BufferSet();
	for (const v of newValidators) {
		newBLSKeys.add(v.blsKey);
		allBLSKeysSet.add(v.blsKey);
		const currentValidator = currentValidatorsMap.get(v.blsKey);
		// if new validator does not exist, add to the blsKeysUpdate and validatorsUpdate
		if (!currentValidator) {
			blsKeysUpdate.push(v.blsKey);
			validatorsUpdate.push(v);
			continue;
		}
		// if currentValidator exist, but bft weight is different, then add to the validatorsUpdate
		if (currentValidator.bftWeight !== v.bftWeight) {
			validatorsUpdate.push(v);
		}
	}

	for (const v of currentValidators) {
		// We add the validator to the validatorsUpdate (with 0 BFT weight) if the BLS key is not in the newValidators.
		if (!newBLSKeys.has(v.blsKey)) {
			validatorsUpdate.push({
				blsKey: v.blsKey,
				bftWeight: BigInt(0),
			});
		}
	}
	blsKeysUpdate.sort((a, b) => a.compare(b));
	validatorsUpdate.sort((a, b) => a.blsKey.compare(b.blsKey));

	const bftWeightsUpdate = validatorsUpdate.map(v => v.bftWeight);
	const allBLSKeys = [...allBLSKeysSet];
	allBLSKeys.sort((a, b) => a.compare(b));

	let bitmap = BigInt(0);
	const bitmapSize = Number(BigInt(allBLSKeys.length + 7) / BigInt(8));
	for (const validator of validatorsUpdate) {
		const index = allBLSKeys.findIndex(k => k.equals(validator.blsKey));
		bitmap |= BigInt(1) << BigInt(index);
	}
	const hexBitmap = bitmap.toString(16);
	// prepend zero if length is odd number
	let bftWeightsUpdateBitmap = Buffer.from(
		hexBitmap.length % 2 === 0 ? hexBitmap : `0${hexBitmap}`,
		'hex',
	);
	// if the result does not have enough length, pad zero to the left
	if (bftWeightsUpdateBitmap.length < bitmapSize) {
		bftWeightsUpdateBitmap = Buffer.concat([
			Buffer.alloc(bitmapSize - bftWeightsUpdateBitmap.length),
			bftWeightsUpdateBitmap,
		]);
	}

	return {
		blsKeysUpdate,
		bftWeightsUpdate,
		bftWeightsUpdateBitmap,
	};
};
