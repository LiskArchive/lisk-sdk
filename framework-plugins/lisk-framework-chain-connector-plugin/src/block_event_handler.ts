/*
 * Copyright Â© 2024 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import {
	chain,
	CCMProcessedResult,
	CcmProcessedEventData,
	CcmSendSuccessEventData,
	LastCertificate,
	MODULE_NAME_INTEROPERABILITY,
	codec,
	ChainAccount,
	EMPTY_BYTES,
} from 'lisk-sdk';
import { ChainAPIClient } from './chain_api_client';
import { ChainConnectorDB } from './db';
import { BlockHeader, Logger, ModuleMetadata } from './types';
import { CCM_PROCESSED, CCM_SEND_SUCCESS } from './constants';
import { CCUHandler } from './ccu_handler';

export interface NewBlockHandlerConfig {
	registrationHeight: number;
	ownChainID: Buffer;
	receivingChainID: Buffer;
	maxCCUSize: number;
	ccuFee: string;
	isSaveCCU: boolean;
	ccuSaveLimit: number;
}

interface NewBlockHandlerInitArgs {
	logger: Logger;
	db: ChainConnectorDB;
	sendingChainAPIClient: ChainAPIClient;
	receivingChainAPIClient: ChainAPIClient;
}

interface Data {
	readonly blockHeader: chain.BlockHeaderJSON;
}
type FinalizedHeightInfo = { inboxSize: number; lastCertificateHeight: number };

export class BlockEventHandler {
	private readonly _ownChainID!: Buffer;
	private readonly _ccuHandler!: CCUHandler;
	private _db!: ChainConnectorDB;
	private _logger!: Logger;
	private _sendingChainAPIClient!: ChainAPIClient;
	private _receivingChainAPIClient!: ChainAPIClient;
	private _lastCertificate!: LastCertificate;
	private _interoperabilityMetadata!: ModuleMetadata;
	private _heightToDeleteIndex!: Map<number, FinalizedHeightInfo>;
	private readonly _ccuFrequency!: number;
	private _receivingChainFinalizedHeight!: number;
	private readonly _ccuSaveLimit: number;
	private readonly _receivingChainID: Buffer;
	private _isReceivingChainRegistered = false;

	public constructor(config: NewBlockHandlerConfig) {
		this._ownChainID = config.ownChainID;
		this._ccuSaveLimit = config.ccuSaveLimit;
		this._receivingChainFinalizedHeight = 0;
		this._receivingChainID = config.receivingChainID;
		this._ccuHandler = new CCUHandler({
			maxCCUSize: config.maxCCUSize,
			ownChainID: config.ownChainID,
			receivingChainID: config.receivingChainID,
			registrationHeight: config.registrationHeight,
			ccuFee: config.ccuFee,
			isSaveCCU: config.isSaveCCU,
		});
	}

	public async load(args: NewBlockHandlerInitArgs) {
		this._logger = args.logger;
		this._db = args.db;
		this._sendingChainAPIClient = args.sendingChainAPIClient;
		this._receivingChainAPIClient = args.receivingChainAPIClient;
		await this._receivingChainAPIClient.connect();
		this._interoperabilityMetadata = await this._sendingChainAPIClient.getMetadataByModuleName(
			MODULE_NAME_INTEROPERABILITY,
		);
		this._lastCertificate = {
			height: 0,
			stateRoot: EMPTY_BYTES,
			timestamp: 0,
			validatorsHash: EMPTY_BYTES,
		};

		this._ccuHandler.load({
			db: args.db,
			lastCertificate: this._lastCertificate,
			logger: args.logger,
			receivingChainAPIClient: args.receivingChainAPIClient,
			sendingChainAPIClient: args.sendingChainAPIClient,
			interoperabilityMetadata: this._interoperabilityMetadata,
		});
		this._heightToDeleteIndex = new Map();
		// On a new block start with CCU creation process
		this._sendingChainAPIClient.subscribe(
			'chain_newBlock',
			async (data?: Record<string, unknown>) => this.handleNewBlock(data),
		);

		this._sendingChainAPIClient.subscribe(
			'chain_deleteBlock',
			async (data?: Record<string, unknown>) => this._deleteBlockHandler(data),
		);
	}

	public async handleNewBlock(data?: Record<string, unknown>) {
		const { blockHeader: receivedBlock } = data as unknown as Data;

		const newBlockHeader = chain.BlockHeader.fromJSON(receivedBlock).toObject();
		let chainAccount: ChainAccount | undefined;
		// Save blockHeader, aggregateCommit, validatorsData and cross chain messages if any.
		const nodeInfo = await this._sendingChainAPIClient.getNodeInfo();
		// Fetch last certificate from the receiving chain and update the _lastCertificate
		try {
			// eslint-disable-next-line no-console
			console.log('this._ownChainID>>>', this._ownChainID);

			chainAccount = await this._receivingChainAPIClient.getChainAccount(this._ownChainID);
			// eslint-disable-next-line no-console
			console.log('chainAccount>>>>>>>>>>>', chainAccount);
		} catch (error) {
			// If receivingChainAPIClient is not ready then still save data on new block
			await this._saveOnNewBlock(newBlockHeader);
			await this._initializeReceivingChainClient();
			this._logger.error(
				{ err: error as Error },
				'Error occurred while accessing receivingChainAPIClient but all data is saved on new block.',
			);

			return;
		}

		// If sending chain is not registered with the receiving chain then only save data on new block and exit
		if (!chainAccount || (chainAccount && !chainAccount.lastCertificate)) {
			this._logger.info(
				'Sending chain is not registered to the receiving chain yet and has no chain data.',
			);
			try {
				await this._saveOnNewBlock(newBlockHeader);
			} catch (error) {
				this._logger.error(
					{ err: error as Error },
					'Error occurred while saving data on new block.',
				);
			}

			return;
		}

		try {
			this._lastCertificate = chainAccount.lastCertificate;
			await this._saveOnNewBlock(newBlockHeader);

			const numOfBlocksSinceLastCertificate = newBlockHeader.height - this._lastCertificate.height;
			if (nodeInfo.syncing || this._ccuFrequency > numOfBlocksSinceLastCertificate) {
				this._logger.debug(
					{
						syncing: nodeInfo.syncing,
						ccuFrequency: this._ccuFrequency,
						nextPossibleCCUHeight: this._ccuFrequency - numOfBlocksSinceLastCertificate,
					},
					'No attempt to create CCU either due to ccuFrequency or the node is syncing',
				);

				return;
			}
		} catch (error) {
			this._logger.error(error, 'Failed while saving on new block');
			return;
		}

		// Check if receiving chain is registered yet or not
		if (!this._isReceivingChainRegistered) {
			const receivingChainAccount = await this._sendingChainAPIClient.getChainAccount(
				this._receivingChainID,
			);
			if (!receivingChainAccount) {
				return;
			}
			this._isReceivingChainRegistered = true;
		}
		// When all the relevant data is saved successfully then try to create CCU
		const computedCCUParams = await this._ccuHandler.computeCCU();

		if (computedCCUParams) {
			try {
				await this._ccuHandler.submitCCU(computedCCUParams.ccuParams);
				// If CCU was sent successfully then save the lastSentCCM if any
				// TODO: Add function to check on the receiving chain whether last sent CCM was accepted or not
				if (computedCCUParams.lastCCMToBeSent) {
					await this._db.setLastSentCCM(computedCCUParams.lastCCMToBeSent);
				}
			} catch (error) {
				this._logger.info(
					{ err: error },
					`Error occured while submitting CCU for the blockHeader at height: ${newBlockHeader.height}`,
				);
			}
		}
	}

	public async _saveOnNewBlock(newBlockHeader: BlockHeader) {
		// Save block header if a new block header arrives
		await this._db.saveOnNewBlock(newBlockHeader);

		this._logger.info('Saved data on new block');
		// Check for events if any and store them
		const events = await this._sendingChainAPIClient.getEvents(newBlockHeader.height);

		const ccmsFromEvents = [];
		// eslint-disable-next-line no-restricted-syntax, no-labels
		ccmInEventsCheck: if (events && events.length > 0) {
			const ccmSendSuccessEvents = events.filter(
				eventAttr =>
					eventAttr.name === CCM_SEND_SUCCESS && eventAttr.module === MODULE_NAME_INTEROPERABILITY,
			);

			const ccmProcessedEvents = events.filter(
				eventAttr =>
					eventAttr.name === CCM_PROCESSED && eventAttr.module === MODULE_NAME_INTEROPERABILITY,
			);

			if (ccmSendSuccessEvents.length === 0 && ccmProcessedEvents.length === 0) {
				// If there are no CCMs present in the events for the height then skip CCM saving part
				// eslint-disable-next-line no-labels
				break ccmInEventsCheck;
			}

			// Save ccm send success events
			if (ccmSendSuccessEvents.length > 0) {
				const ccmSendSuccessEventInfo = this._interoperabilityMetadata.events.filter(
					e => e.name === CCM_SEND_SUCCESS,
				);

				if (!ccmSendSuccessEventInfo?.[0]?.data) {
					throw new Error('No schema found for "ccmSendSuccess" event data.');
				}

				for (const e of ccmSendSuccessEvents) {
					const eventData = codec.decode<CcmSendSuccessEventData>(
						ccmSendSuccessEventInfo[0].data,
						Buffer.from(e.data, 'hex'),
					);
					ccmsFromEvents.push(eventData.ccm);
				}
			}
			// Save ccm processed events based on CCMProcessedResult FORWARDED = 1
			if (ccmProcessedEvents.length > 0) {
				const ccmProcessedEventInfo = this._interoperabilityMetadata.events.filter(
					e => e.name === CCM_PROCESSED,
				);

				if (!ccmProcessedEventInfo?.[0]?.data) {
					throw new Error('No schema found for "ccmProcessed" event data.');
				}

				for (const e of ccmProcessedEvents) {
					const eventData = codec.decode<CcmProcessedEventData>(
						ccmProcessedEventInfo[0].data,
						Buffer.from(e.data, 'hex'),
					);
					if (eventData.result === CCMProcessedResult.FORWARDED) {
						ccmsFromEvents.push(eventData.ccm);
					}
				}
			}
		}
		this._logger.info('Collected CCMs');

		await this._db.setCCMsByHeight(
			ccmsFromEvents.map(ccm => ({ ...ccm, height: newBlockHeader.height })),
			newBlockHeader.height,
		);

		this._logger.info('Set CCMs');

		const validatorsData = await this._sendingChainAPIClient.getBFTParametersAtHeight(
			newBlockHeader.height,
		);

		this._logger.info('get getBFTParametersAtHeight');

		await this._db.setValidatorsDataByHash(validatorsData.validatorsHash, validatorsData);
		this._logger.info('set getBFTParametersAtHeight');
	}

	private async _initializeReceivingChainClient() {
		try {
			await this._receivingChainAPIClient.connect();
			this._receivingChainAPIClient.subscribe(
				'chain_newBlock',
				async (data?: Record<string, unknown>) => this._newBlockReceivingChainHandler(data),
			);
		} catch (error) {
			this._logger.error(
				error,
				'Not able to connect to receivingChainAPIClient. Trying again on next new block.',
			);
		}
	}

	private async _newBlockReceivingChainHandler(_?: Record<string, unknown>) {
		try {
			const { finalizedHeight } = await this._receivingChainAPIClient.getNodeInfo();
			this._receivingChainFinalizedHeight = finalizedHeight;
			const { inbox } = await this._receivingChainAPIClient.getChannelAccount(this._ownChainID);
			if (!inbox) {
				throw new Error('No channel data available on receiving chain.');
			}
			const chainAccount = await this._receivingChainAPIClient.getChainAccount(this._ownChainID);
			if (!chainAccount?.lastCertificate) {
				throw new Error('No chain data available on receiving chain.');
			}
			this._heightToDeleteIndex.set(finalizedHeight, {
				inboxSize: inbox.size,
				lastCertificateHeight: chainAccount.lastCertificate?.height,
			});
			await this._cleanup();
		} catch (error) {
			this._logger.debug(
				error,
				'No Channel or Chain Data: Sending chain is not registered yet on receiving chain.',
			);
		}
	}

	private async _cleanup() {
		// Delete CCUs
		// When given -1 then there is no limit
		if (this._ccuSaveLimit !== -1) {
			const listOfCCUs = await this._db.getListOfCCUs();
			if (listOfCCUs.length > this._ccuSaveLimit) {
				await this._db.setListOfCCUs(
					// Takes the last ccuSaveLimit elements
					listOfCCUs.slice(-this._ccuSaveLimit),
				);
			}
			let finalizedInfoAtHeight = this._heightToDeleteIndex.get(
				this._receivingChainFinalizedHeight,
			);
			if (!finalizedInfoAtHeight) {
				for (let i = 1; i < this._heightToDeleteIndex.size; i += 1) {
					if (this._heightToDeleteIndex.get(this._receivingChainFinalizedHeight - i)) {
						finalizedInfoAtHeight = this._heightToDeleteIndex.get(
							this._receivingChainFinalizedHeight - i,
						);
						break;
					}
				}
			}

			if (this._lastCertificate.height > 0) {
				// Delete CCMs
				await this._db.deleteCCMsBetweenHeight(
					1,
					finalizedInfoAtHeight ? finalizedInfoAtHeight.inboxSize : 0,
				);

				// Delete blockHeaders
				await this._db.deleteBlockHeadersBetweenHeight(
					1,
					finalizedInfoAtHeight ? finalizedInfoAtHeight.lastCertificateHeight : 0,
				);

				// Delete aggregateCommits
				await this._db.deleteAggregateCommitsBetweenHeight(
					1,
					finalizedInfoAtHeight ? finalizedInfoAtHeight.lastCertificateHeight : 0,
				);
				// Delete validatorsHashPreimage
			}

			// Delete info less than finalized height
			this._heightToDeleteIndex.forEach((_, key) => {
				if (key < this._receivingChainFinalizedHeight) {
					this._heightToDeleteIndex.delete(key);
				}
			});
		}
	}

	private async _deleteBlockHandler(data?: Record<string, unknown>) {
		const { blockHeader: receivedBlock } = data as unknown as Data;

		const deletedBlockHeader = chain.BlockHeader.fromJSON(receivedBlock).toObject();

		// Delete ccmEvents for the height of blockHeader
		await this._db.deleteCCMsByHeight(deletedBlockHeader.height);
		await this._db.deleteBlockHeaderByHeight(deletedBlockHeader.height);
		await this._db.deleteAggregateCommitByHeight(deletedBlockHeader.height);
	}
}
