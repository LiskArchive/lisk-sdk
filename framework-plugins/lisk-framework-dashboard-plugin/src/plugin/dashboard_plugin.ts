/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { objects } from '@liskhq/lisk-utils';
import {
	ActionsDefinition,
	BasePlugin,
	BaseChannel,
	EventsDefinition,
	PluginInfo,
	SchemaWithDefault,
} from 'lisk-framework';
import * as express from 'express';
import { join } from 'path';
import { Server } from 'http';
import * as defaults from './defaults';
import { dashboardPluginOptions } from './types';

// eslint-disable-next-line
const packageJSON = require('../../package.json');

export class DashboardPlugin extends BasePlugin {
	private _options!: dashboardPluginOptions;
	private _server!: Server;

	// eslint-disable-next-line @typescript-eslint/class-literal-property-style
	public static get alias(): string {
		return 'dashboard';
	}

	// eslint-disable-next-line @typescript-eslint/class-literal-property-style
	public static get info(): PluginInfo {
		return {
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
			author: packageJSON.author,
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
			version: packageJSON.version,
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
			name: packageJSON.name,
		};
	}

	// eslint-disable-next-line class-methods-use-this
	public get defaults(): SchemaWithDefault {
		return defaults.config;
	}

	// eslint-disable-next-line class-methods-use-this
	public get events(): EventsDefinition {
		return [];
	}

	// eslint-disable-next-line class-methods-use-this
	public get actions(): ActionsDefinition {
		return {};
	}

	// eslint-disable-next-line @typescript-eslint/require-await, class-methods-use-this
	public async load(_channel: BaseChannel): Promise<void> {
		this._options = objects.mergeDeep(
			{},
			defaults.config.default,
			this._options,
		) as dashboardPluginOptions;
		const app = express();
		app.use(express.static(join(__dirname, '../../build')));
		this._server = app.listen(this._options.port, this._options.host);
	}

	// eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-empty-function
	public async unload(): Promise<void> {
		await new Promise((resolve, reject) => {
			this._server.close(err => {
				if (err) {
					reject(err);
					return;
				}
				resolve();
			});
		});
	}
}
