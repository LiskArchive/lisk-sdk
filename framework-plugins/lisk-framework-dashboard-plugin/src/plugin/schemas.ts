/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const configSchema = {
	$id: '#/plugins/liskDashboard/config',
	type: 'object',
	properties: {
		applicationName: {
			type: 'string',
			description: 'Application name to be shown near Logo',
		},
		applicationUrl: {
			type: 'string',
			format: 'uri',
			description: 'URL to connect',
		},
		port: {
			type: 'integer',
			minimum: 1,
			maximum: 65535,
		},
		host: {
			type: 'string',
			format: 'ip',
		},
	},
	required: [],
	default: {
		applicationUrl: 'ws://127.0.0.1:7887/rpc-ws',
		port: 4005,
		host: '127.0.0.1',
		applicationName: 'Lisk',
	},
};
