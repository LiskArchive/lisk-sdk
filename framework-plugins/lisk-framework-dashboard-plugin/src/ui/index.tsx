/* eslint-disable import/extensions */
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import './index.scss';
import { App } from './app';

ReactDOM.render(
	<React.StrictMode>
		<App />
	</React.StrictMode>,
	document.getElementById('root'),
);
