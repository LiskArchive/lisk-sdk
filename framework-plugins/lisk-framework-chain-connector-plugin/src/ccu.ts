/*
 * Copyright Â© 2022 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import {
	BFTHeights,
	ActiveValidator,
	codec,
	certificateSchema,
	CrossChainUpdateTransactionParams,
	apiClient,
	LastCertificate,
	ccmSchema,
	tree,
	CCMsg,
	OutboxRootWitness,
	db as liskDB,
	crossChainUpdateTransactionParams,
} from 'lisk-sdk';
import { getNextCertificateFromAggregateCommits } from './certificate_generation';
import { calculateActiveValidatorsUpdate } from './active_validators_update';
import { ChainConnectorStore } from './db';
import { BlockHeader, InboxUpdate, CrossChainMessagesFromEvents } from './types';

export class Ccu {
	private readonly _chainConnectorPluginDB!: liskDB.Database;
	private readonly _sidechainChainConnectorStore: ChainConnectorStore;
	private readonly _sidechainAPIClient!: apiClient.APIClient;
	private readonly _ownChainID!: Buffer;
	private readonly _lastCertificate!: LastCertificate;

	public constructor(
		store: ChainConnectorStore,
		chainApiClient: apiClient.APIClient,
		ownChainID: Buffer,
		certificate: LastCertificate,
	) {
		this._sidechainChainConnectorStore = store;
		this._sidechainAPIClient = chainApiClient;
		this._ownChainID = ownChainID;
		this._lastCertificate = certificate;
	}

	public async calculateCCUParams(): Promise<void> {
		const blockHeaders = await this._sidechainChainConnectorStore.getBlockHeaders();
		const aggregateCommits = await this._sidechainChainConnectorStore.getAggregateCommits();
		const validatorsHashPreimages =
			await this._sidechainChainConnectorStore.getValidatorsHashPreImage();
		const bftHeights = await this._sidechainAPIClient.invoke<BFTHeights>('consensus_getBFTHeights');
		// Calculate certificate
		const certificate = getNextCertificateFromAggregateCommits(
			blockHeaders,
			aggregateCommits,
			validatorsHashPreimages,
			bftHeights,
			this._lastCertificate,
		);
		if (!certificate) {
			throw new Error('No certificate found in the ccu creation process.');
		}
		// Calculate activeValidatorsUpdate
		let activeValidatorsUpdate: ActiveValidator[] = [];
		let certificateThreshold = BigInt(0);
		const blockHeader = blockHeaders.find(header => header.height === certificate.height);
		if (!blockHeader) {
			throw new Error('No block header found for the given certificate height.');
		}
		if (!this._lastCertificate.validatorsHash.equals(certificate.validatorsHash)) {
			const validatorsUpdateResult = calculateActiveValidatorsUpdate(
				blockHeader,
				validatorsHashPreimages,
				this._lastCertificate,
			);
			activeValidatorsUpdate = validatorsUpdateResult.activeValidatorsUpdate;
			certificateThreshold = validatorsUpdateResult.certificateThreshold;
		}
		// Calculate inboxUpdate
		const inboxUpdates = await this._calculateInboxUpdate(blockHeader);

		// Now create CCUs for all the inboxUpdates handling full or partial updates
		for (const inboxUpdate of inboxUpdates) {
			const serializedCertificate = codec.encode(certificateSchema, certificate);
			const serializedCCUParams = codec.encode(crossChainUpdateTransactionParams, {
				sendingChainID: this._ownChainID,
				activeValidatorsUpdate,
				certificate: serializedCertificate,
				certificateThreshold,
				inboxUpdate,
			} as CrossChainUpdateTransactionParams);
			await this._createCCU(serializedCCUParams);
		}
	}

	private async _calculateInboxUpdate(newBlockHeader: BlockHeader): Promise<InboxUpdate[]> {
		const ccmsFromEvents = await this._sidechainChainConnectorStore.getCrossChainMessages();
		const ccmsListOfList = this._getListOfCCMs(ccmsFromEvents, newBlockHeader);

		if (ccmsListOfList.length === 1) {
			const ccms = ccmsListOfList[0];
			const crossChainMessages = ccms.map(ccm => codec.encode(ccmSchema, ccm));
			// Take the inclusion proof of the last ccm height
			const { inclusionProof } = ccms[ccms.length - 1];

			return [
				{
					crossChainMessages,
					messageWitnessHashes: [],
					outboxRootWitness: inclusionProof,
				},
			];
		}

		// Calculate list of inboxUpdates to be sent by multiple CCUs
		const inboxUpdates = [];
		for (const subList of ccmsListOfList) {
			const crossChainMessages = subList.map(ccm => codec.encode(ccmSchema, ccm));
			// Take the inclusion proof of the last ccm height
			const { inclusionProof } = subList[subList.length - 1];

			// Calculate message witnesses

			const merkleTree = new tree.MerkleTree({ db: this._chainConnectorPluginDB });
			for (const ccm of crossChainMessages) {
				await merkleTree.append(ccm);
			}
			const messageWitnessHashes = await merkleTree.generateRightWitness(crossChainMessages.length);

			inboxUpdates.push({
				crossChainMessages,
				messageWitnessHashes,
				outboxRootWitness: inclusionProof,
			});
		}

		return inboxUpdates;
	}

	private _getListOfCCMs(
		_crossChainMessages: CrossChainMessagesFromEvents[],
		_newBlockHeader: BlockHeader,
	): {
		ccm: CCMsg;
		height: number;
		inclusionProof: OutboxRootWitness;
	}[][] {
		return [[]];
	}

	// eslint-disable-next-line @typescript-eslint/no-empty-function
	private async _createCCU(_serializedCCUParams: Buffer) {}
}
