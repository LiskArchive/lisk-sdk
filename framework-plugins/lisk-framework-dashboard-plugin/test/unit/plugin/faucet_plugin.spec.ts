/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { DashboardPlugin } from '../../../src/plugin';

describe('DashboardPlugin', () => {
	describe('defaults', () => {
		it('should return valid config schema with default options', () => {
			// eslint-disable-nextline
			const plugin = new DashboardPlugin({} as any);

			expect(plugin.defaults).toMatchSnapshot();
		});
	});

	describe('constructor', () => {
		it('should load default config', () => {
			// eslint-disable-nextline
			const plugin = new DashboardPlugin({} as any);

			expect(plugin.options).toMatchSnapshot();
		});
	});
});
