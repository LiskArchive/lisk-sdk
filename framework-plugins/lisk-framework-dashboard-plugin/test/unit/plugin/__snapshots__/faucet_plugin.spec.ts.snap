// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardPlugin constructor should load default config 1`] = `
Object {
  "applicationUrl": "ws://localhost:8080",
  "host": "127.0.0.1",
  "port": 4000,
}
`;

exports[`DashboardPlugin defaults should return valid config schema with default options 1`] = `
Object {
  "$id": "#/plugins/lisk-dashboard/config",
  "default": Object {
    "applicationUrl": "ws://localhost:8080",
    "host": "127.0.0.1",
    "port": 4000,
  },
  "properties": Object {
    "applicationUrl": Object {
      "description": "URL to connect",
      "format": "uri",
      "type": "string",
    },
    "host": Object {
      "format": "ip",
      "type": "string",
    },
    "port": Object {
      "maximum": 65535,
      "minimum": 1,
      "type": "integer",
    },
  },
  "required": Array [],
  "type": "object",
}
`;
